sql/src/main/java/io/crate/execution/jobs/MergeBucketsReceiver.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.jobs;
+
+import io.crate.Streamer;
+import io.crate.data.Bucket;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * A {@link PageBucketReceiver} which receives paged buckets from upstreams
+ * and forwards the merged bucket results to the consumers for further processing.
```

Comment:
I think the most important note here would be that it will wait for the current page from all upstreams before it continues with the next page